### PR TITLE
JAMES-3775 Guice packaging for RSpamd

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]

**Review by chibenwa — COMMENTED** [review]
Please register webadmin routes too in the example

**Comment by vttranlina** [comment]
Remark: 
- James is not supporting the extension for `Task json dto`.
- We cannot declare Guice inject for `webadmin-dto` by normal way:
```
    @Named(DTOModuleInjections.WEBADMIN_DTO)
    @ProvidesIntoSet
    public AdditionalInformationDTOModule<? extends TaskExecutionDetails.AdditionalInformation, ? extends AdditionalInformationDTO> webAdminFeedSpamAdditionalInformation() {
     ...
    }
```
(Guice throws an error)

-> Don't allow the use of the routes for now.

**Comment by Arsnael** [comment]
What do you mean don't allow the use of the routes for now???

File: examples/metrics-graphite/pom.xml [hunk 1]
```diff
@@ -25,6 +25,7 @@
         <version>3.8.0-SNAPSHOT</version>
     </parent>
 
+
```
Comment (Arsnael) [inline examples/metrics-graphite/pom.xml:28]: No need and unrelated

**Comment by vttranlina** [comment]
> What do you mean don't allow the use of the routes for now???

yes, because we can't start James. 

@chibenwa opinion
```
That's not possible yet...
- 1. for now cheat: don't allow the use of the routes for now.
 - 2. Write on the mailing list about your need.
 - 3. We will define and contribute such an extension mechanism 
```

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/module/RSpamDModule.java [hunk 2]
```diff
@@ -49,4 +57,11 @@ protected void configure() {
     public AdditionalInformationDTOModule<? extends TaskExecutionDetails.AdditionalInformation, ? extends AdditionalInformationDTO> webAdminFeedSpamAdditionalInformation() {
         return FeedSpamToRSpamDTaskAdditionalInformationDTO.SERIALIZATION_MODULE;
     }
+
+    @Provides
+    @Singleton
+    public RSpamDHttpClient rSpamDHttpClient(PropertiesProvider propertiesProvider) throws ConfigurationException, FileNotFoundException {
+        RSpamDClientConfiguration rSpamDClientConfiguration = RSpamDClientConfiguration.from(propertiesProvider.getConfiguration("rspamd"));
+        return new RSpamDHttpClient(rSpamDClientConfiguration);
+    }
```
Comment (quantranhong1999) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/module/RSpamDModule.java:40]: Should we create bean from `RSpamDClientConfiguration` level? (we inject it into the client https://github.com/apache/james-project/blob/eb2178a9269851cf4d15eed98669faa172cedd61/third-party/rspamd/src/main/java/org/apache/james/rspamd/client/RSpamDHttpClient.java).

And we still can set `RSpamDHttpClient` instance to singleton if we want.

File: third-party/rspamd/docker-compose.yml [hunk 3]
```diff
@@ -3,11 +3,29 @@ version: '3'
 services:
 
   james:
+    depends_on:
+      - rspamd
     image: apache/james:memory-latest
     container_name: james
     hostname: james.local
     volumes:
+      - /home/tungtv/workplace/2_JAMES/ENV/tmail-backend-memory/keystore:/root/conf/keystore
```
Comment (quantranhong1999) [inline third-party/rspamd/docker-compose.yml:12]: Please share the `keystore` file.

**Comment by quantranhong1999** [comment]
```
Cleaning up the workspace
/home/jenkins/jenkins-agent/workspace/james: Read-only file system
```

Likely related to the latest Jenkins cleanup commit.

**Comment by vttranlina** [comment]
We already have manually `docker-compose-distributed.yml`
Should we still have a deployment test?

File: third-party/rspamd/sample-configuration/extensions.properties [hunk 4]
```diff
@@ -1 +1,2 @@
-guice.extension.module=org.apache.james.rspamd.module.RSpamDModule
\ No newline at end of file
+guice.extension.module=org.apache.james.rspamd.module.RSpamDModule
+guice.extension.task=org.apache.james.rspamd.module.RSpamDTaskExtensionModule
```
Comment (quantranhong1999) [inline third-party/rspamd/sample-configuration/extensions.properties:2]: Q: Do we still need `guice.extension.module=org.apache.james.rspamd.module.RSpamDModule` when we declared addtional route in `webadmin.properties` and DTO modules in `RSpamDTaskExtensionModule` ?

Comment (vttranlina) [reply third-party/rspamd/sample-configuration/extensions.properties:2]: yes.

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by Arsnael** [comment]
@vttranlina needs a rebase :)

**Comment by vttranlina** [comment]
> @vttranlina needs a rebase :)

I tried to rebase, but this pr depends on many other pr. 
It has a bit complex to rebase & resolve. So, I will rebase this pr later, after more pr has been merged. It will be better

**Comment by chibenwa** [comment]
Conflict spotted

**Pull request merged**